### PR TITLE
Add System.Memory reference

### PR DESCRIPTION
**Pull request by agocke**
Allows us to replace some unsafe code with Span<T>. I anticipate more complex
uses in the future.

Here are the benchmarks.

**Before**

```
BenchmarkDotNet=v0.11.1, OS=Windows 10.0.17134.345 (1803/April2018Update/Redstone4)
Intel Core i7-7700 CPU 3.60GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=2.2.200-preview-009431
  [Host] : .NET Core 2.1.6 (CoreCLR 4.6.27019.06, CoreFX 4.6.27019.05), 64bit RyuJIT
  Core   : .NET Core 2.1.6 (CoreCLR 4.6.27019.06, CoreFX 4.6.27019.05), 64bit RyuJIT

Job=Core  Runtime=Core  Server=True

                Method |      Mean |     Error |    StdDev |    Median |     Gen 0 | Allocated |
---------------------- |----------:|----------:|----------:|----------:|----------:|----------:|
               Parsing |  96.86 ms |  2.568 ms |  7.571 ms |  96.89 ms |         - |   4.25 MB |
 CompileMethodsAndEmit | 688.64 ms | 13.659 ms | 38.971 ms | 681.50 ms | 1000.0000 |  43.97 MB |
     SerializeMetadata | 231.56 ms | 10.722 ms | 31.613 ms | 244.91 ms |         - |  18.06 MB |
```

**After**

```
BenchmarkDotNet=v0.11.1, OS=Windows 10.0.17134.345 (1803/April2018Update/Redstone4)
Intel Core i7-7700 CPU 3.60GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=2.2.200-preview-009431
  [Host] : .NET Core 2.1.6 (CoreCLR 4.6.27019.06, CoreFX 4.6.27019.05), 64bit RyuJIT
  Core   : .NET Core 2.1.6 (CoreCLR 4.6.27019.06, CoreFX 4.6.27019.05), 64bit RyuJIT

Job=Core  Runtime=Core  Server=True

                Method |      Mean |     Error |    StdDev |    Median |     Gen 0 | Allocated |
---------------------- |----------:|----------:|----------:|----------:|----------:|----------:|
               Parsing |  95.71 ms |  2.381 ms |  7.021 ms |  95.44 ms |         - |   3.59 MB |
 CompileMethodsAndEmit | 708.76 ms | 19.100 ms | 55.108 ms | 707.51 ms | 1000.0000 |  44.02 MB |
     SerializeMetadata | 199.04 ms |  8.754 ms | 24.834 ms | 191.37 ms |         - |  18.06 MB |
```

There does not appear to be any statistically significant difference.

### CODE REVIEW

File: src/Compilers/Core/Portable/InternalUtilities/StringTable.cs [hunk 1]
```diff
@@ -730,25 +736,7 @@ internal static unsafe bool TextEqualsASCII(string text, byte* ascii, int length
...
-                {
-                    return false;
-                }
-                s++;
-            }
-
-            return true;
-        }
+        internal static bool TextEquals(string array, ReadOnlySpan<char> text)
+            => text.Equals(array.AsSpan(), StringComparison.Ordinal);
```
Comment (Therzok) [inline src/Compilers/Core/Portable/InternalUtilities/StringTable.cs]: I think this should be SequenceEquals.

File: src/Compilers/Core/Portable/InternalUtilities/StringTable.cs [hunk 2]
```diff
@@ -705,21 +711,21 @@ internal static bool TextEquals(string array, StringBuilder text)
...
         }
 
-        internal static unsafe bool TextEqualsASCII(string text, byte* ascii, int length)
+        internal static bool TextEqualsASCII(string text, ReadOnlySpan<byte> ascii)
         {
 #if DEBUG
-            for (var i = 0; i < length; i++)
+            for (var i = 0; i < ascii.Length; i++)
             {
                 Debug.Assert((ascii[i] & 0x80) == 0, "The byte* input to this method must be valid ASCII.");
```
Comment (jcouv) [inline src/Compilers/Core/Portable/InternalUtilities/StringTable.cs]: >byte* [](start = 58, length = 5)

This message is out-of-date

File: src/Compilers/Core/Portable/InternalUtilities/StringTable.cs (hunk 1 above)
Comment (jcouv) [inline src/Compilers/Core/Portable/InternalUtilities/StringTable.cs]: >text.Equals [](start = 15, length = 11)

I didn't understand how this `.Equals` works. 
`ReadOnlySpan.Equals` seems to throw an exception ([source](https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System/ReadOnlySpan.cs#L63))

File: src/Compilers/Core/Portable/InternalUtilities/TextKeyedCache.cs [hunk 3]
```diff
@@ -118,7 +118,7 @@ internal T FindItem(char[] chars, int start, int len, int hashCode)
 
             if (text != null && localSlot.HashCode == hashCode)
             {
-                if (StringTable.TextEquals(text, chars, start, len))
+                if (StringTable.TextEquals(text, chars.AsSpan(start, len)))
```
Comment (jcouv) [inline src/Compilers/Core/Portable/InternalUtilities/TextKeyedCache.cs]: >AsSpan [](start = 55, length = 6)

`AsReadOnlySpan`? (also below)

**Comment by jcouv** [comment]
I assume the benchmarks in the description were for a previous iteration with a broader change.

File: src/Setup/DevDivInsertionFiles/DevDivInsertionFiles.csproj [hunk 4]
```diff
@@ -44,10 +44,13 @@
...
     <ExpectedDependency Include="Microsoft.DiaSymReader" DevEnvLib="lib/netstandard1.1/Microsoft.DiaSymReader.dll"/>
     <ExpectedDependency Include="Microsoft.CodeAnalysis.Elfie" DevEnvLib="lib/net46/Microsoft.CodeAnalysis.Elfie.dll"/>
+    <ExpectedDependency Include="System.Buffers" DevEnvLib="lib/netstandard2.0/System.Buffers.dll" />
     <ExpectedDependency Include="System.Collections.Immutable" DevEnvLib="lib/netstandard2.0/System.Collections.Immutable.dll"/>
     <ExpectedDependency Include="System.Reflection.Metadata" DevEnvLib="lib/netstandard2.0/System.Reflection.Metadata.dll"/>
+    <ExpectedDependency Include="System.Memory" DevEnvLib="lib/netstandard2.0/System.Memory.dll" />
     <ExpectedDependency Include="System.Runtime.CompilerServices.Unsafe" DevEnvLib="lib/netstandard2.0/System.Runtime.CompilerServices.Unsafe.dll"/>
     <ExpectedDependency Include="System.Text.Encoding.CodePages" DevEnvLib="lib/netstandard2.0/System.Text.Encoding.CodePages.dll"/>
+    <ExpectedDependency Include="System.Numerics.Vectors" DevEnvLib="lib/netstandard2.0/System.Numerics.Vectors.dll" />
```
Comment (jcouv) [inline src/Setup/DevDivInsertionFiles/DevDivInsertionFiles.csproj]: >System.Numerics.Vectors [](start = 33, length = 23)

Does this relate to System.Memory?

File: src/Compilers/Core/CodeAnalysisTest/AnalyzerFileReferenceTests.cs [hunk 5]
```diff
@@ -223,6 +223,8 @@ public class TestAnalyzer : DiagnosticAnalyzer
 
             dir.CopyFile(typeof(System.Reflection.Metadata.MetadataReader).Assembly.Location);
             dir.CopyFile(typeof(AppDomainUtils).Assembly.Location);
+            dir.CopyFile(typeof(Memory<byte>).Assembly.Location);
```
Comment (jcouv) [inline src/Compilers/Core/CodeAnalysisTest/AnalyzerFileReferenceTests.cs]: >byte [](start = 39, length = 4)

nit: the `byte` is probably not needed: `typeof(Memory<>)` (here and in `TestHelpers.cs`)

File: src/Setup/DevDivInsertionFiles/DevDivInsertionFiles.csproj (hunk 4 above)
Comment (agocke) [reply src/Setup/DevDivInsertionFiles/DevDivInsertionFiles.csproj]: Dependent reference.

File: src/Compilers/Core/Portable/InternalUtilities/TextKeyedCache.cs (hunk 3 above)
Comment (agocke) [reply src/Compilers/Core/Portable/InternalUtilities/TextKeyedCache.cs]: There's an implicit conversion, doesn't matter.

File: src/Compilers/Core/Portable/InternalUtilities/StringTable.cs (hunk 1 above)
Comment (agocke) [reply src/Compilers/Core/Portable/InternalUtilities/StringTable.cs]: https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System/MemoryExtensions.Fast.cs#L37

Comment (jcouv) [reply src/Compilers/Core/Portable/InternalUtilities/StringTable.cs]: Thanks!

**Pull request merged**